platform/linux-generic/odp_classification.c

```diff
@@ -1555,56 +1555,45 @@ static inline void pmr_debug_print(pmr_t *pmr, cos_t *cos)
 
 /*
  * Match a PMR chain with a Packet and return matching CoS
- * This function gets called recursively to check the chained PMR Term value
- * with the packet.
+ * This function performs a depth-first search in the CoS tree.
  */
-static cos_t *match_pmr_cos(cos_t *cos, const uint8_t *pkt_addr, pmr_t *pmr,
-			    odp_packet_hdr_t *hdr)
+static cos_t *match_pmr_cos(cos_t *cos, const uint8_t *pkt_addr, odp_packet_hdr_t *hdr)
 {
-	uint32_t i, num_rule;
+	while (1) {
+		uint32_t i, num_rule = odp_atomic_load_u32(&cos->num_rule);
 
-	if (cos == NULL || pmr == NULL)
-		return NULL;
-
-	if (!cos->valid)
-		return NULL;
+		for (i = 0; i < num_rule; i++) {
+			pmr_t *pmr = cos->pmr[i];
 
-	if (verify_pmr(pmr, pkt_addr, hdr)) {
-		/* PMR matched */
-		pmr_debug_print(pmr, cos);
+			if (odp_unlikely(!cos->linked_cos[i]->valid))
+				continue;
 
-		if (cos->stats_enable)
-			odp_atomic_inc_u64(&cos->stats.packets);
+			if (verify_pmr(pmr, pkt_addr, hdr)) {
+				/* PMR matched */
 
-		hdr->p.input_flags.cls_mark = 0;
-		if (pmr->mark) {
-			hdr->p.input_flags.cls_mark = 1;
-			hdr->cls_mark = pmr->mark;
-		}
+				cos = cos->linked_cos[i];
```

Comment:
Current implementation has race between updating PMR/linked_cos and reading those here. We'll need to fix that later (with atomic read of pmr+linked_cos). However, I'd suggest to read cos->linked_cos only once here to minimize issues where linked_cos->valid and cos=linked_cos would refer to different linked_cos pointers.

Reply:
v3.